libs/libcommon/src/libcommon/orchestrator.py

```diff
@@ -0,0 +1,757 @@
+                    cache_status.cache_is_error_to_retry[artifact_state.id] = artifact_state
+                    continue
+
+                # was created with an obsolete version of the job runner?
+                if artifact_state.cache_state.is_job_runner_obsolete():
+                    cache_status.cache_is_job_runner_obsolete[artifact_state.id] = artifact_state
+                    continue
+
+                # has a different git revision from the dataset current revision?
+                if artifact_state.cache_state.is_git_revision_different_from(self.revision):
+                    cache_status.cache_has_different_git_revision[artifact_state.id] = artifact_state
+                    continue
+
+                # ok
+                cache_status.up_to_date[artifact_state.id] = artifact_state
+
+        return cache_status
+
+    def get_queue_status(self) -> QueueStatus:
+        processing_steps = (
+            self.processing_graph.get_first_processing_steps()
+            if self.only_first_processing_steps
+            else self.processing_graph.get_topologically_ordered_processing_steps()
+        )
+        return QueueStatus(
+            in_process={
+                artifact_state.id
+                for processing_step in processing_steps
+                for artifact_state in self._get_artifact_states_for_step(processing_step)
+                if artifact_state.job_state.is_in_process
+            }
+        )
+
+    def _create_plan(self) -> None:
+        pending_jobs_to_delete_df = self.pending_jobs_df.copy()
+        job_infos_to_create: List[JobInfo] = []
+        artifact_states = (
+            list(self.cache_status.cache_is_empty.values())
+            + list(self.cache_status.cache_is_error_to_retry.values())
+            + list(self.cache_status.cache_is_outdated_by_parent.values())
+            + list(self.cache_status.cache_is_job_runner_obsolete.values())
+            + list(self.cache_status.cache_has_different_git_revision.values())
+        )
+        for artifact_state in artifact_states:
+            valid_pending_jobs_df = artifact_state.job_state.valid_pending_jobs_df
+            if valid_pending_jobs_df.empty:
+                job_infos_to_create.append(
+                    {
+                        "job_id": "not used",
+                        "type": artifact_state.processing_step.job_type,
+                        "params": {
+                            "dataset": self.dataset,
+                            "revision": self.revision,
+                            "config": artifact_state.config,
+                            "split": artifact_state.split,
+                        },
+                        "priority": self.priority,
+                    }
+                )
+            else:
+                pending_jobs_to_delete_df.drop(valid_pending_jobs_df.index, inplace=True)
+        # Better keep this order: delete, then create
```

Comment:
There is precisely similar logic in AfterJobPlan post_init, could it be factored in a method?

Reply:
Hmm it's only four lines, I think it's not a problem to keep them duplicated.